### PR TITLE
[distributions] clip sigmoid to prevent transforms return inf/nan values

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR addresses some numerical issues of Sigmoid/StickBreakingTransform, where these transforms give +-inf when the unconstrained values move to +-20 areas.

For example, with
```
t = torch.distributions.SigmoidTransform()
x = torch.tensor(20.)
t.inv(t(x)), t.log_abs_det_jacobian(x, t(x))
```
current behaviour the inverse will return `inf` and logdet return `-inf` while this PR makes it to `15.9424` and `-15.9424`.

And for
```
t = torch.distributions.StickBreakingTransform()
x = torch.tensor([20., 20.])
t.inv(t(x)), t.log_abs_det_jacobian(x, t(x))
```
current value is `(inf, nan)` and `-inf` for logdet, while this PR makes it `[16.6355, 71.3942]` and `-47.8272` for logdet.

Although these finite values are wrong and seems unavoidable, it is better than returning `inf` or `nan` in my opinion. This is useful in HMC where despite that the grad will be zero when the unconstrained parameter moves to unstable area (due to clipping), velocity variable will force the parameter move to another area which by chance can move the parameter out of unstable area. But inf/nan can be useful to stop doing inference early. So the changes in this PR might be inappropriate.

I also fix some small issues of `_Simplex` and `_RealVector` constraints where batch shape of the input is not respected when checking validation.

### CODE REVIEW

**Comment by ezyang** [comment]
Once you get review from other distributions maintainers, use pytorchbot to request a merge

**Comment by fehiepsi** [comment]
Oppss, I forgot to cc reviewers. @neerajprad Could you please help me review this?

**Review by neerajprad — COMMENTED** [review]
The math looks fine to me. Since other users have also brought up numerical issues with transforms like SigmoidTransform, this seems like a reasonable change to me. I think that we should surely not return `NaN`s from `log_abs_det_jacobian`, but I would like to get other opinions on whether there might be unintended effects when clamping to not return `inf`/`-inf` from transforms.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py]: This is mainly to save the last transform computation, where we don't need to recompute `y`. It can help numerical stability *somehow* by using an exact `y` instead of an approximated `y` through transforms.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py]: Thanks! I'll revert this with a drop for the unnecessary `expand` operator.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py]: Yup, when we use `transforms`, it is better to compute log_det in term of `x` (to eliminate numerical precision while getting `y` from `x`). Thanks for providing a reference! I'll see if it really helps and gets back on this.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py]: Yes.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py]: Maybe we can discuss about this in #19802 first to let it merged first, then I'll make corresponding changes in this PR later.

Comment (neerajprad) [reply torch/distributions/transforms.py]: Sure. Now that I am looking at that PR again, I think we should probably just define a clamping function to clamp to `(0, 1)` and call that in both `_call` and `_inverse`, because users might be using the inverse transform with values that can be 0 or 1.

Comment (fehiepsi) [reply torch/distributions/transforms.py]: Yeah, agree.

**Comment by fehiepsi** [comment]
@neerajprad #19802 is merged so we can keep discussing the changes in this PR.

File: torch/distributions/transforms.py
Comment (neerajprad) [inline torch/distributions/transforms.py]: I think we can use this identity but replace the sigmoid and clamping by using `F.logsigmoid` instead:

```python
detJ = (-x + F.logsigmoid(x) + y[..., :-1].log()).sum(-1)
```

Comment (fehiepsi) [reply torch/distributions/transforms.py]: Thanks @neerajprad ! I just address it. :)

**Review by neerajprad — APPROVED** [review]
Looks great!

**Pull request closed without merging**
